### PR TITLE
feat: implements almost all of the i32 operations

### LINKED ISSUE #87: incorrect implementation of i32.div_s
The current implementation treats `div_s` and `div_u` as the same operation. However, as per the specification, `div_s` uses a 'signed' interpretation of the values:

https://webassembly.github.io/spec/core/exec/numerics.html#aux-signed

this has an impact for numbers where the most significant bit is '1'. 

To illustrate:

**unsigned**

```
(module
  (func $test (param i32) (param i32) (result i32)
   (get_local 0)
   (get_local 1)
   (i32.div_u)
  )
  (export "test" (func $test))
)
```
```
instance.exports.test(0xffffff00, 0x2);
```
```
2147483520
```

**signed**

```
(module
  (func $test (param i32) (param i32) (result i32)
   (get_local 0)
   (get_local 1)
   (i32.div_s)
  )
  (export "test" (func $test))
)
```
```
instance.exports.test(0xffffff00, 0x2);
```
```
-128
```

### PR DESCRIPTION
**Pull request by ColinEberhardt**
progress towards #97 

 - [x] i32.add
 - [x] i32.sub
 - [x] i32.mul
 - [x] i32.div_u
 - [ ] i32.div_s - needs to throw 'integer overflow' exception
 - [x] i32.or
 - [x] i32.xor
 - [x] i32.and
 - [x] i32.rem_u
 - [x] i32.rem_s
 - [x] i32.shl
 - [x] i32.shr_u
 - [x] i32.shr_s
 - [x] i32.rotl
 - [x] i32.rotr
 - [x] i32.clz
 - [x] i32.ctz
 - [x] i32.popcnt
 - [x] i32.eq
 - [x] i32.eqz
 - [x] i32.ine
 - [x] i32.ilt_u
 - [x] i32.ilt_s
 - [x] i32.igt_u
 - [x] i32.igt_s
 - [x] i32.ile_u
 - [x] i32.ile_s
 - [x] i32.ige_u
 - [x] i32.ige_s

fixes #87 

I'm basically running through all of our integer operations with a goal of getting them fully spec compliant and closing off #97 

I've taken the official webassembly test suite and used a few regexes to convert it into chai / mocha test. Once #76 is implemented, these tests can be deleted.

Regarding this specific change, `i32` is now a class that implements our `NumberInterface` interface, allowing implementation of some of the specific features we need to match WebAssembly integer maths. With this PR we now support signed / unsigned division, multiplication uses Long to ensure we don't exceed JS numeric precision, and various runtime exceptions are being thrown (as per spec)

### CODE REVIEW

**Comment by ColinEberhardt** [comment]
updated to add:
 - [x] i32.and
 - [x] i32.rem_u
 - [x] i32.rem_s

**Comment by ColinEberhardt** [comment]
updated to add:
 - [x] i32.shl
 - [x] i32.shr_u
 - [x] i32.shr_s

File: test/interpreter/fixtures/integer-operations/exec.tjs [hunk 1]
```diff
@@ -79,52 +79,52 @@ it("should conform to the wasm test suite", () => {
   equals(m.exports.div_u(11, 5), 2);
   equals(m.exports.div_u(17, 7), 2);
 
-  // throws(() => m.exports.rem_s(1, 0), "$4");
```
Comment (xtuc) [inline test/interpreter/fixtures/integer-operations/exec.tjs]: Could we remove it instead of commenting it?

**Comment by ColinEberhardt** [comment]
- [x] rotl
 - [x] rotr

File: src/interpreter/runtime/values/i32.js [hunk 2]
```diff
@@ -1,14 +1,167 @@
...
 
-export class i32 extends BaseNumber {}
+// the specification describes the conversion from unsigned to signed
+// https://webassembly.github.io/spec/core/exec/numerics.html#aux-signed
+// this function performs the inverse
+const toUnsigned = a => (a < 0 ? a + Math.pow(2, bits) : a);
 
-export function createValueFromAST(value: number): StackLocal {
-  value = (value | 0) % Math.pow(2, bits);
+export class i32 implements Number<i32> {
```
Comment (xtuc) [inline src/interpreter/runtime/values/i32.js]: You probably need to rebase, I had to rename the interface.

**Comment by xtuc** [comment]
Looks nice! I will review it in detail on my desktop tommorow.

File: src/interpreter/runtime/values/i32.js (hunk 2 above)
Comment (ColinEberhardt) [reply src/interpreter/runtime/values/i32.js]: Ah - good spot, I've turned Flow back on again, and fixed the issues this PR introduced :-)

**Comment by ColinEberhardt** [comment]
- [x] clz
 - [x] ctz

File: src/interpreter/runtime/values/number.js [hunk 3]
```diff
@@ -69,6 +70,92 @@ export class BaseNumber implements NumberInterface<BaseNumber> {
     return new BaseNumber(-this._value);
   }
 
+  /*eslint-disable no-unused-vars*/
```
Comment (xtuc) [inline src/interpreter/runtime/values/number.js]: Just FYI I commented the operand somewhere in a PR, like:
```js
rem_s(/*operand: BaseNumber*/): BaseNumber {
    throw new RuntimeError("Unsupported operation");
}
```

Comment (ColinEberhardt) [reply src/interpreter/runtime/values/number.js]: Ah, good point - for some reason I thought doing that would result in Flow errors, but it does not :-)

File: src/interpreter/kernel/instruction/unop.js [hunk 4]
```diff
@@ -19,6 +19,18 @@ function unop(
...
+      return createValue(c.clz());
+
+    case "ctz":
+      return createValue(c.ctz());
+
+    case "popcnt":
+      return createValue(c.popcnt());
+
+    case "eqz":
+      return createValue(c.eqz());
```
Comment (xtuc) [inline src/interpreter/kernel/instruction/unop.js]: By the way, why is the value called `c`? value sounds more explicit to me.

File: src/interpreter/runtime/values/i32.js [hunk 5]
```diff
@@ -1,14 +1,271 @@
...
+        : -this._value
+    );
+  }
+
+  toNumber(): number {
+    return this._value;
+  }
+
+  isTrue(): boolean {
+    return this._value == 1;
```
Comment (xtuc) [inline src/interpreter/runtime/values/i32.js]: We can use triple equal here. I'm not sure about this tbh, numeric types are not booleans, either the name is misleading or we could inline the `=== 1` in the callers?

File: test/interpreter/fixtures/integer-operations/module.wast [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+   (i32.ge_s)
+  )
+  (export "ge_s" (func $ge_s))
+  (func $ge_u (param i32) (param i32) (result i32)
+   (get_local 0)
+   (get_local 1)
+   (i32.ge_u)
+  )
+  (export "ge_u" (func $ge_u))
+)
```
Comment (xtuc) [inline test/interpreter/fixtures/integer-operations/module.wast]: I really like that, the spec is way more testable this way, but way less maintainable.

File: src/interpreter/kernel/instruction/unop.js (hunk 4 above)
Comment (ColinEberhardt) [reply src/interpreter/kernel/instruction/unop.js]: good point - I'll change that.

**Pull request merged**